compiler/circle-mpqsolver/src/bisection/Evaluator.cpp

```diff
@@ -0,0 +1,155 @@
+ * limitations under the License.
+ */
+
+#include "Evaluator.h"
+
+#include <luci_interpreter/Interpreter.h>
+
+#include <dio_hdf5/HDF5Importer.h>
+
+#include <cmath>
+#include <iostream>
+
+using namespace mpqsolver::bisection;
+
+using Shape = std::vector<loco::Dimension>;
+
+namespace
+{
+
+using namespace luci;
+
+template <typename NodeT> size_t get_tensor_size(const NodeT *node)
+{
+  uint32_t tensor_size = loco::size(node->dtype());
+  for (uint32_t i = 0; i < node->rank(); ++i)
+    tensor_size *= node->dim(i).value();
+  return tensor_size;
+}
+
+DatasetOutput compute_outputs(const luci::Module *module, const std::string &h5file)
+{
+  dio::hdf5::HDF5Importer importer{h5file};
+  importer.importGroup("value");
+
+  bool is_raw_data = importer.isRawData();
+
+  const auto num_records = importer.numData();
+  if (num_records == 0)
+    throw std::runtime_error("The input data file does not contain any record.");
+  const auto input_nodes = loco::input_nodes(module->graph());
+  const auto num_inputs = input_nodes.size();
+
+  DatasetOutput dataset_output;
+
+  // Create interpreter.
+  luci_interpreter::Interpreter interpreter(module);
+  for (int32_t record_idx = 0; record_idx < num_records; record_idx++)
+  {
+    if (num_inputs != importer.numInputs(record_idx))
+      throw std::runtime_error("Wrong number of inputs.");
+    for (int32_t input_idx = 0; input_idx < num_inputs; input_idx++)
+    {
+      const auto *input_node = loco::must_cast<const luci::CircleInput *>(input_nodes[input_idx]);
+      assert(input_node->index() == input_idx);
+
+      std::vector<char> input_data(get_tensor_size(input_node));
+
+      if (!is_raw_data)
+      {
+        loco::DataType dtype;
+        Shape shape;
+        importer.readTensor(record_idx, input_idx, &dtype, &shape, input_data.data());
+      }
+      else
+      {
+        // Skip type/shape check for raw data
+        importer.readTensor(record_idx, input_idx, input_data.data());
+      }
+
+      interpreter.writeInputTensor(input_node, input_data.data(), input_data.size());
+    }
+
+    interpreter.interpret();
+
+    NNOutput nn_output;
+
+    // Get output.
+    const auto output_nodes = loco::output_nodes(module->graph());
+    for (int i = 0; i < module->graph()->outputs()->size(); i++)
+    {
+      const auto *output_node = loco::must_cast<const luci::CircleOutput *>(output_nodes[i]);
+      ElementaryOutput output_data(get_tensor_size(output_node));
+      interpreter.readOutputTensor(output_node, output_data.data(), output_data.size());
+      // output
+      nn_output.push_back(output_data);
+    }
+    dataset_output.push_back(nn_output);
+  }
+
+  return dataset_output;
+}
+
+} // namespace
+
+DatasetEvaluator::DatasetEvaluator(const luci::Module *ref_module, const std::string &h5file)
+  : _ref_module(ref_module), _h5file(h5file)
+{
+  _ref_output = compute_outputs(_ref_module, _h5file);
+}
+
+float DatasetEvaluator::evaluate(const luci::Module *trgt_fq_module) const
```

Comment:
Currenlty evaluate only returns Mean Absolute Error. Other metrics may be added later.

Could you make the metric configurable?

Reply:
yes, i can. I'll add parameter to the constructor.

Reply:
One way to support configurable metric is to make a `Metric` class which has interface. `MetricPrinter.h` in circle-eval-diff may be helpful.

Reply:
> One way to support configurable metric is to make a `Metric` class which has interface. `MetricPrinter.h` in circle-eval-diff may be helpful.

Ahh.. I see. Seems like it's more preferable.